extension/icu/icu-timebucket.cpp

```diff
@@ -0,0 +1,526 @@
+#include "duckdb/common/operator/cast_operators.hpp"
+#include "duckdb/common/types/interval.hpp"
+#include "duckdb/common/types/time.hpp"
+#include "duckdb/common/types/timestamp.hpp"
+#include "duckdb/common/types/value.hpp"
+#include "duckdb/common/vector_operations/binary_executor.hpp"
+#include "duckdb/common/vector_operations/ternary_executor.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/parser/parsed_data/create_scalar_function_info.hpp"
+#include "include/icu-datefunc.hpp"
+
+namespace duckdb {
+
+struct ICUTimeBucket : public ICUDateFunc {
+
+	// Use 2000-01-03 00:00:00 (Monday) as origin when bucket_width is days, hours, ... for TimescaleDB compatibility
+	// There are 10959 days between 1970-01-01 and 2000-01-03
+	constexpr static const int64_t DEFAULT_ORIGIN_MICROS_1 = 10959 * Interval::MICROS_PER_DAY;
+	// Use 2000-01-01 as origin when bucket_width is months, years, ... for TimescaleDB compatibility
+	// There are 10957 days between 1970-01-01 and 2000-01-01
+	constexpr static const int64_t DEFAULT_ORIGIN_MICROS_2 = 10957 * Interval::MICROS_PER_DAY;
+	// There are 360 months between 1970-01-01 and 2000-01-01
+	constexpr static const int64_t DEFAULT_ORIGIN_MONTHS = 360;
+
+	enum struct BucketWidthType { LESS_THAN_DAYS, MORE_THAN_MONTHS, UNCLASSIFIED };
+
+	static inline BucketWidthType ClassifyBucketWidth(const interval_t bucket_width) {
+		if (bucket_width.months == 0 && Interval::GetMicro(bucket_width) > 0) {
+			return BucketWidthType::LESS_THAN_DAYS;
+		} else if (bucket_width.months > 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+			return BucketWidthType::MORE_THAN_MONTHS;
+		} else {
+			return BucketWidthType::UNCLASSIFIED;
+		}
+	}
+
+	static inline BucketWidthType ClassifyBucketWidthWithThrowingError(const interval_t bucket_width) {
+		if (bucket_width.months == 0) {
+			int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+			if (bucket_width_micros <= 0) {
+				throw NotImplementedException("Period must be greater than 0");
+			}
+			return BucketWidthType::LESS_THAN_DAYS;
+		} else if (bucket_width.months != 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+			if (bucket_width.months < 0) {
+				throw NotImplementedException("Period must be greater than 0");
+			}
+			return BucketWidthType::MORE_THAN_MONTHS;
+		} else {
+			throw NotImplementedException("Month intervals cannot have day or time component");
+		}
+	}
+
+	static inline int32_t EpochMonths(timestamp_t ts, icu::Calendar *calendar) {
+		SetTime(calendar, ts);
+		return (ExtractField(calendar, UCAL_YEAR) - 1970) * 12 + ExtractField(calendar, UCAL_MONTH);
+	}
+
+	static inline timestamp_t WidthLessThanDaysCommon(int64_t bucket_width_micros, timestamp_t ts, timestamp_t origin,
+	                                                  icu::Calendar *calendar) {
+		int64_t ts_micros = Interval::GetMicro(Sub(calendar, ts, origin));
+		int64_t result_micros = (ts_micros / bucket_width_micros) * bucket_width_micros;
```

Comment:
I think this is not correct because with DST a day is not always 24 hours long!  For intervals less than days, you need to just use the basic µs delta `ts.value - origin.value`. But you also need a _third_ common routine for anything with days, and then you need the full interval value so you can do a kind of long division between the result of `Sub` and that interval.

Reply:
Introduced a new routine for bucket_width that can be converted to days.

Reply:
I am having a problem with `interval_t` to `int64_t` conversion and vice versa.

`int64_t ts_micros = Interval::GetMicro(Sub(calendar, ts, origin))` does not compute the correct value because `Interval::GetMicro` always treats a day (even DST boundary day) as 24 hours.
`Interval::FromMicro(result_micros)` in the return statement also calculates incorrectly by the same reason.
It looks like the two calculation errors cancel each other out and it almost works!

So rewiting this function needs corrections in two lines.
```C++
static inline timestamp_t WidthConvertibleToMicrosCommon(int64_t bucket_width_micros, const timestamp_t ts,
                                                         const timestamp_t origin, icu::Calendar *calendar) {
	int64_t ts_micros = SubtractOperatorOverflowCheck::Operation<int64_t, int64_t, int64_t>(ts.value, origin.value);
	int64_t result_micros = (ts_micros / bucket_width_micros) * bucket_width_micros;
	if (ts_micros < 0 && ts_micros % bucket_width_micros != 0) {
		result_micros =
		    SubtractOperatorOverflowCheck::Operation<int64_t, int64_t, int64_t>(result_micros, bucket_width_micros);
	}

	return Add(calendar, origin, interval_t {0, 0, result_micros});
}
```
looks reasoneble, but it failed because `ICUCalendarAdd::Operation` in `Add` seems not to calculate the correct result when interval's microsecond is large.

For example
```sql
D select '2000-01-01 00:00:00+00'::timestamptz + '2400 hours'::interval;
┌───────────────────────────────────────────────────────────────────────────────────────────────┐
│ (CAST('2000-01-01 00:00:00+00' AS TIMESTAMP WITH TIME ZONE) + CAST('2400 hours' AS INTERVAL)) │
│                                   timestamp with time zone                                    │
├───────────────────────────────────────────────────────────────────────────────────────────────┤
│ 2000-01-01 13:54:25.408+00                                                                    │
└───────────────────────────────────────────────────────────────────────────────────────────────┘
```

Reply:
When you are using sub-day intervals, then you don't need to worry about day boundaries. That's why I said you should split out a separate case for when there are days in the `interval_t`. That is also what ICU does if I am not mistaken, so that is the way the world works.

I'm surprised by the error in `Add` - ICU does have resolution issues with large values, but they are typically measured in millennia, not days. I'll try to have a look, but my time is becoming limited.  its a very simple repro though and should be filed as a separate issue.

Reply:
Thanks for the fix #5888 !

Reply:
> When you are using sub-day intervals, then you don't need to worry about day boundaries. 

`Interval::GetMicro` I used does not use ICU and treats the 23-hour or 25-hour DST boundary day as a 24-hour day.
So `int64_t ts_micros = Interval::GetMicro(Sub(calendar, ts, origin))` calculates with ±1 hour error when there is a DST boundary between `ts` and `origin`.
I don't need to worry about DST boundaries when getting epoch microseconds from `timestamp_t` directly without going through `interval_t`.

Reply:
Yes and I think this is correct. After all, my date add fix shows that ICU does this in `Americas/Los_Angeles`:

```sql
set Calendar = 'gregorian';
set TimeZone = 'America/Los_Angeles';
select '1999-12-31 16:00:00-08'::timestamptz + interval 2400 hours
-- 2000-04-09 17:00:00-07
```

In other words, it adds 2400 hours, not 100 days. Which is what I would expect.

Reply:
Think you can resolve this conversation